src/PMCSimulation.jl

```diff
@@ -0,0 +1,409 @@
+"""
+    PMCSimulation
+Holds the state and the results of a projector quantum Monte Carlo (PMC) simulation.
+Initialise with [`init(::ProjectorMonteCarloProblem)`](@ref) and solve with
+[`solve!(::PMCSimulation)`](@ref).
+
+Obtain the results of a simulation `sm` as a DataFrame with `DataFrame(sm)`.
+
+See also [`state_vectors`](@ref), [`single_states`](@ref),
+[`ProjectorMonteCarloProblem`](@ref), [`init`](@ref), [`solve!`](@ref).
+"""
+mutable struct PMCSimulation
+    problem::ProjectorMonteCarloProblem
+    algorithm # currently only FCIQMC() is implemented
+    state::ReplicaState
+    report::Report
+    modified::Bool
+    aborted::Bool
+    success::Bool
+    message::String
+    elapsed_time::Float64
+end
+
+function _set_up_v(dv::AbstractDVec, copy_vectors, _...)
+    # we are not changing type, style or initiator because an AbstractDVec is passed
+    if copy_vectors
+        return deepcopy(dv)
+    else
+        return dv
+    end
+end
+
+function _set_up_v(fdv::FrozenDVec, _, style, initiator, threading)
+    # we are allocating new memory
+    if threading
+        v = PDVec(fdv; style, initiator)
+    elseif initiator isa NonInitiator
+        v = DVec(fdv; style)
+    else
+        v = InitiatorDVec(fdv; style, initiator)
+    end
+    return v
+end
+
+function PMCSimulation(problem::ProjectorMonteCarloProblem; copy_vectors=true)
+    @unpack algorithm, hamiltonian, starting_vectors, style, threading, simulation_plan,
+        replica_strategy, shift_strategy, initial_shift_parameters,
+        reporting_strategy, post_step_strategy, time_step_strategy,
+        maxlength, metadata, initiator, random_seed, spectral_strategy = problem
+
+    n_replicas = num_replicas(replica_strategy)
+
+    # seed the random number generator
+    if !isnothing(random_seed)
+        Random.seed!(random_seed + hash(RMPI.mpi_rank()))
+    end
+
+    # set up the starting vectors and shift parameters
+    if length(starting_vectors) == 1 && n_replicas > 1
+        vectors = ntuple(n_replicas) do i
+            v = _set_up_v(only(starting_vectors), copy_vectors, style, initiator, threading)
+            sizehint!(v, maxlength)
```

Comment:
```suggestion
```
I'd advise against doing this. `sizehint!` has a pretty minor effect (and only when the `Dict` in the vector is growing, so first few steps), and sometimes you want to set `maxlength` to something large (like `typemax(Int)`) to prevent Rimu from complaining.